File: utils/abi_types.ts
import {
  ABIAddressType,
  ABIArrayDynamicType,
  ABIArrayStaticType,
  ABIBoolType,
  ABIByteType,
  ABITupleType,
  ABIType,
  ABIUintType
} from "algosdk";

export const RequestInfoType = new ABITupleType([
  new ABIArrayStaticType(new ABIByteType, 32), // request_id
  new ABIUintType(64), // voting_contract
  new ABIUintType(64), // request_round
  new ABIUintType(8), // request_status
  new ABIUintType(64), // total_stake
  new ABIArrayStaticType(new ABIByteType, 32), // sha512_256(Concat(<Request Sender>, <Key>))
  new ABIBoolType(), //is_history
  new ABIUintType(64), // requester_algo_fee
  new ABIUintType(64), // total_votes
  new ABIUintType(64), // total_votes_refunded

]);

export const ProposalsEntryType = new ABITupleType([
  new ABIArrayStaticType(new ABIByteType, 32),  // vote_hash
  new ABIUintType(64), // vote_count
  new ABIUintType(64), // stake_count
  new ABIUintType(64), // vote_round
  new ABIUintType(64), // rewards_payed_out
  new ABIAddressType(), // requester
  new ABIBoolType(), //is_history
]);

export const LocalHistoryType = new ABITupleType([
  new ABIArrayStaticType(new ABIByteType, 32),  // key_hash
  ProposalsEntryType // proposal_entry
]);

export const StakeTupleType = new ABITupleType([
  new ABIUintType(64), // round
  new ABIUintType(64) // total_stake
]);

export const BoxReference = new ABITupleType([
  new ABIArrayDynamicType(new ABIByteType), // name
  new ABIUintType(64) //app_id
]);

export const BoxReferenceList = new ABIArrayDynamicType(BoxReference);

export const StakeArrayType = new ABIArrayStaticType(StakeTupleType,2);

export const ResponseBodyType = new ABITupleType([
  ABIType.from("byte[32]"), //request_id
  ABIType.from("address"), //requester_address
  ABIType.from("byte[]"), // oracle return value
  ABIType.from("byte[]"), // user data
  ABIType.from("uint32"), // error code
  ABIType.from("uint64") // source failure bitmap
]);

export const ResponseBodyBytesType = new ABIArrayDynamicType(new ABIByteType);

export const SourceSpecType = new ABITupleType([
  new ABIUintType(32), //source_id #TODO we still need to decide the mapping of these
  new ABIArrayDynamicType(new ABIArrayDynamicType(new ABIByteType())), // source args...this is usually the JSON path
  new ABIUintType(64) //max age of the data in seconds 
]);

export const RequestArgsType = new ABITupleType([
  new ABIArrayDynamicType(SourceSpecType),
  new ABIUintType(32), 
  new ABIArrayDynamicType(new ABIByteType)
]);

export const DestinationType = new ABITupleType([
  new ABIUintType(64),  //app id of the destination 
  new ABIArrayDynamicType(new ABIByteType()) //method signature of the method in the destination app
]); 

export const SubscriptionType = new ABITupleType([
  new ABIUintType(16),
  new ABIUintType(16)
]);

export function create_source_arr(sourceID: number, source_args: Uint8Array[], maxAgeSeconds: number)
{
  return [sourceID, source_args, maxAgeSeconds];
}

export function create_source_spec(sourceID: number, source_args: Uint8Array[], maxAgeSeconds: number)
{
  return SourceSpecType.encode([sourceID, source_args, maxAgeSeconds]);
}

export function create_request_args(sourceSpecArray: [][], aggregationType: number, userData: Uint8Array)
{
  return RequestArgsType.encode([sourceSpecArray, aggregationType, userData ]);
}

export function create_destination_arg(destinationAppID: number, destinationMethodSignature: Uint8Array)
{
  return DestinationType.encode([destinationAppID, destinationMethodSignature]);
}

File: template_content/protocol/utils/abi_types.py
from pyteal import *
from algosdk import abi as algosdkAbi
from typing import Literal as L

class StakeHistoryTuple(abi.NamedTuple):
    round: abi.Field[abi.Uint64]
    total_stake: abi.Field[abi.Uint64]

StakeHistoryTupleAlgoSDK = algosdkAbi.TupleType([
        algosdkAbi.UintType(64), # round
        algosdkAbi.UintType(64), # total_stake
    ])

class BoxRef(abi.NamedTuple):
    name: abi.Field[abi.DynamicBytes]
    app_id: abi.Field[abi.Uint64]

class ResponseBody(abi.NamedTuple):
    request_id: abi.Field[abi.StaticBytes[L[32]]]
    requester_address: abi.Field[abi.Address]
    oracle_return_value: abi.Field[abi.DynamicArray[abi.Byte]]
    user_data: abi.Field[abi.DynamicArray[abi.Byte]]
    error_code: abi.Field[abi.Uint32]
    source_failures: abi.Field[abi.Uint64]

class RequestInfo(abi.NamedTuple):
    request_id: abi.Field[abi.StaticBytes[L[32]]]
    voting_contract: abi.Field[abi.Uint64]
    request_round: abi.Field[abi.Uint64]
    request_status: abi.Field[abi.Uint8]
    total_stake: abi.Field[abi.Uint64]
    key_hash: abi.Field[abi.StaticBytes[L[32]]] # sha512_256(Concat(<Request Sender>, <Key>))
    # is_history must be the same index as in ProposalsEntry
    is_history: abi.Field[abi.Bool] 
    requester_algo_fee: abi.Field[abi.Uint64]
    total_votes: abi.Field[abi.Uint64]
    total_votes_refunded: abi.Field[abi.Uint64]

class ProposalsEntry(abi.NamedTuple):
    vote_hash: abi.Field[abi.StaticBytes[L[32]]]
    vote_count: abi.Field[abi.Uint64]
    stake_count: abi.Field[abi.Uint64]
    vote_round: abi.Field[abi.Uint64]
    rewards_payed_out: abi.Field[abi.Uint64]
    requester: abi.Field[abi.Address]
    # is_history must be the same index as in RequestInfo
    is_history: abi.Field[abi.Bool]

class LocalHistoryEntry(abi.NamedTuple):
    key_hash: abi.Field[abi.StaticBytes[L[32]]] # sha512_256(Concat(<Request Sender>, <Key>))
    proposal_entry: abi.Field[ProposalsEntry]

class SourceSpec(abi.NamedTuple):
    source_id: abi.Field[abi.Uint32]
    source_arg_list: abi.Field[abi.DynamicBytes]
    max_age: abi.Field[abi.Uint64]

class RequestSpec(abi.NamedTuple):
    source_specs: abi.Field[abi.DynamicArray[SourceSpec]]
    aggregation: abi.Field[abi.Uint32]
    user_data: abi.Field[abi.DynamicBytes]

class DestinationSpec(abi.NamedTuple):
    destination_id: abi.Field[abi.Uint64]
    destination_method: abi.Field[abi.DynamicBytes]
